File: ui/src/ports/connect_api.ts
// Role: port (driving) — what the frontend may ask about connecting: what this machine
// offers, which far end the window is on, and starting a different one.

import type {
  ConnectAnswer,
  ConnectQuestion,
  Connectable,
  Connected,
  LaunchRequest,
  ProfileId,
  SavedConnections,
  SetUp,
} from '../protocol';

/**
 * A far end that asks a question nobody is listening for is told nobody answered, which the
 * backend reads as a refusal.
 */
export interface ConnectListener {
  onQuestion?(question: ConnectQuestion): Promise<ConnectAnswer>;
  onProgress?(said: string): void;
}

export interface ConnectApi {
  connectable(): Promise<Connectable[]>;
  /**
   * Rejects with a whole spoken sentence when it cannot be started, and the running session
   * is untouched. `origin` is the saved connection this attempt started from, or `null`
   * for a new one.
   */
  use(
    id: ProfileId,
    setUp: SetUp,
    origin: string | null,
    listener?: ConnectListener,
  ): Promise<Connected>;
  /** `null` for a window connected to nothing. */
  connected(): Promise<Connected | null>;

  saved(): Promise<SavedConnections>;

  /** Answers the sentence to say, or rejects with one. */
  saveConnection(name: string): Promise<string>;

  /** Answers the sentence to say, or rejects with one. */
  renameConnection(from: string, to: string): Promise<string>;

  /** Answers the sentence to say, or rejects with one. */
  forgetConnection(name: string): Promise<string>;

  offerToSave(): Promise<boolean>;

  stopOfferingToSave(): Promise<void>;

  /** What `acter --connect <name>` asked for, or `null` for an ordinary launch. */
  requestedAtLaunch(): Promise<LaunchRequest | null>;
}

File: crates/acter-core/src/ports/driving/connect_api.rs
//! Port (driving): what may be asked about connecting — what this machine offers, which
//! far end is behind the window now, and starting one in place of another.

use std::sync::Arc;

use crate::{
    ConnectQuestions, Connectable, Connected, LaunchRequest, ProfileId, SavedConnections, SetUp,
};

/// Every `String` in a result here, `Ok` or `Err`, is a whole sentence to speak.
pub trait ConnectApi: Send + Sync {
    /// Implementations must ask the machine on every call and never cache the list.
    fn connectable(&self) -> Vec<Connectable>;

    /// The new session is built before the old one is dropped, so an `Err` leaves the
    /// current session running and attached.
    ///
    /// Blocks until a person has answered every question, so it must never run on the thread
    /// that delivers the answers.
    ///
    /// `origin` is the saved connection this attempt started from, or `None` for a new one.
    fn use_profile(
        &self,
        id: &ProfileId,
        set_up: SetUp,
        origin: Option<&str>,
        questions: &Arc<dyn ConnectQuestions>,
    ) -> Result<Connected, String>;

    /// `None` means the window is connected to nothing.
    fn connected(&self) -> Option<Connected>;

    fn saved(&self) -> SavedConnections;

    fn save_connection(&self, name: &str) -> Result<String, String>;

    fn rename_connection(&self, from: &str, to: &str) -> Result<String, String>;

    fn forget_connection(&self, name: &str) -> Result<String, String>;

    fn offer_to_save(&self) -> bool;

    fn stop_offering_to_save(&self) -> Result<(), String>;

    /// What `acter --connect <name>` asked for, or `None` for an ordinary launch.
    fn requested_at_launch(&self) -> Option<LaunchRequest>;
}
